File: apps/server/openhive/persistence/tokens.py
"""Token storage — encrypts on write, decrypts on read."""

from __future__ import annotations

import time
from dataclasses import dataclass

from openhive.persistence.crypto import decrypt, encrypt
from openhive.persistence.db import db


@dataclass
class TokenRecord:
    provider_id: str
    access_token: str
    refresh_token: str | None
    expires_at: int | None  # unix seconds
    scope: str | None
    account_label: str | None


def save(record: TokenRecord) -> None:
    now = int(time.time())
    access_enc = encrypt(record.access_token)
    refresh_enc = encrypt(record.refresh_token) if record.refresh_token else None
    with db() as conn:
        conn.execute(
            """
            INSERT INTO oauth_tokens
              (provider_id, access_token, refresh_token, expires_at, scope, account_label,
               created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT(provider_id) DO UPDATE SET
              access_token=excluded.access_token,
              refresh_token=excluded.refresh_token,
              expires_at=excluded.expires_at,
              scope=excluded.scope,
              account_label=excluded.account_label,
              updated_at=excluded.updated_at
            """,
            (
                record.provider_id,
                access_enc,
                refresh_enc,
                record.expires_at,
                record.scope,
                record.account_label,
                now,
                now,
            ),
        )
        conn.commit()


def load(provider_id: str) -> TokenRecord | None:
    with db() as conn:
        row = conn.execute(
            "SELECT * FROM oauth_tokens WHERE provider_id = ?", (provider_id,)
        ).fetchone()
    if not row:
        return None
    return TokenRecord(
        provider_id=row["provider_id"],
        access_token=decrypt(row["access_token"]),
        refresh_token=decrypt(row["refresh_token"]) if row["refresh_token"] else None,
        expires_at=row["expires_at"],
        scope=row["scope"],
        account_label=row["account_label"],
    )


def delete(provider_id: str) -> bool:
    with db() as conn:
        cur = conn.execute("DELETE FROM oauth_tokens WHERE provider_id = ?", (provider_id,))
        conn.commit()
        return cur.rowcount > 0


def list_connected() -> list[str]:
    with db() as conn:
        rows = conn.execute("SELECT provider_id FROM oauth_tokens").fetchall()
    return [r["provider_id"] for r in rows]


def get_account_label(provider_id: str) -> str | None:
    with db() as conn:
        row = conn.execute(
            "SELECT account_label FROM oauth_tokens WHERE provider_id = ?", (provider_id,)
        ).fetchone()
    return row["account_label"] if row else None

File: apps/web/lib/server/tokens.ts
/**
 * OAuth token store — FS-only. Tokens live Fernet-encrypted in
 *   ~/.openhive/oauth.enc.json
 *
 * We hold the full record map in one file and rewrite on save. At single-digit
 * provider counts this is fine; if it ever grows we can shard per-provider.
 */

import fs from 'node:fs'
import path from 'node:path'
import { dataDir } from './paths'
import { decrypt, encrypt } from './crypto'

export interface TokenRecord {
  provider_id: string
  access_token: string
  refresh_token: string | null
  expires_at: number | null
  scope: string | null
  account_label: string | null
  account_id: string | null
}

interface StoredRow {
  provider_id: string
  access_token: string           // encrypted
  refresh_token: string | null   // encrypted
  expires_at: number | null
  scope: string | null
  account_label: string | null
  account_id: string | null
  created_at: number
  updated_at: number
}

function tokensPath(): string {
  return path.join(dataDir(), 'oauth.enc.json')
}

function readAll(): Record<string, StoredRow> {
  const p = tokensPath()
  if (!fs.existsSync(p)) return {}
  try {
    const raw = JSON.parse(fs.readFileSync(p, 'utf8'))
    return (raw && typeof raw === 'object') ? (raw as Record<string, StoredRow>) : {}
  } catch {
    return {}
  }
}

function writeAll(rows: Record<string, StoredRow>): void {
  const p = tokensPath()
  fs.mkdirSync(path.dirname(p), { recursive: true })
  const tmp = `${p}.tmp`
  fs.writeFileSync(tmp, JSON.stringify(rows, null, 2), 'utf8')
  fs.renameSync(tmp, p)
}

export function saveToken(record: TokenRecord): void {
  const rows = readAll()
  const now = Math.floor(Date.now() / 1000)
  const prev = rows[record.provider_id]
  rows[record.provider_id] = {
    provider_id: record.provider_id,
    access_token: encrypt(record.access_token),
    refresh_token: record.refresh_token ? encrypt(record.refresh_token) : null,
    expires_at: record.expires_at,
    scope: record.scope,
    account_label: record.account_label,
    account_id: record.account_id,
    created_at: prev?.created_at ?? now,
    updated_at: now,
  }
  writeAll(rows)
}

export function loadToken(providerId: string): TokenRecord | null {
  const rows = readAll()
  const row = rows[providerId]
  if (!row) return null
  return {
    provider_id: row.provider_id,
    access_token: decrypt(row.access_token),
    refresh_token: row.refresh_token ? decrypt(row.refresh_token) : null,
    expires_at: row.expires_at,
    scope: row.scope,
    account_label: row.account_label,
    account_id: row.account_id,
  }
}

export function deleteToken(providerId: string): boolean {
  const rows = readAll()
  if (!(providerId in rows)) return false
  delete rows[providerId]
  writeAll(rows)
  return true
}

export function listConnected(): string[] {
  return Object.keys(readAll())
}

export function getAccountLabel(providerId: string): string | null {
  const row = readAll()[providerId]
  return row?.account_label ?? null
}
